Feature/add client capability flow

CoralMalachi opened the pull request:
The following pull request adds support for passing Controller Certification
Client Capability Query test.
According to section 5.6.2 in the EasyMesh Test Plan v1.0 [1],
the MCUT (MultiAP Controller Under Test) sends a client capability query
in order to get information about the CTT STA.

The following changes were made :
1.     The Agent subscribes to a message of type :	CLIENT_CAPABILITY_QUERY_MESSAGE.

2.     Add handle method which responsible for handle the CMDU when receive at the agent.

3.     A shell test flow have been added to `test_flows.sh` for the Client Capability Query,
       according to section 5.6.2 in the EasyMesh Test Plan v1.0 [1].

Commit:
Agent: add client capability query handler

According to section 5.6.2 in the EasyMesh Test Plan v1.0 [1],
the MCUT (MultiAP Controller Under Test) sends a client capability query
in order to get information about the CTT STA.

To make sure the query was received by the agent I have added the following changes:

1.	The Agent subscribes to a message of type :
  	CLIENT_CAPABILITY_QUERY_MESSAGE.

2.	Add handle_1905_client_capability_query method which only prints
	"Received Client capability Query message" when the query is :
	received.

Note,`handle_client_capability_query` method should be in the
`son_slave.cpp` which is responsible for the radio.

I have moved this method from `backhaul_manager_thread.cpp` to `son_slave.cpp`,
since it does not make any sense of handling it here because the
backhaul manager forwards messages to son_slaves if not handled them by
itself.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

morantr commented:
Why commit https://github.com/prplfoundation/prplMesh/pull/188/commits/2a1736a574853ce50666b8cac5d67921c550d941 is here?

morantr commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -126,8 +126,8 @@ bool main_thread::init()
-            ieee1905_1::eMessageType::ACK_MESSAGE,
-            ieee1905_1::eMessageType::TOPOLOGY_QUERY_MESSAGE})) {
+            ieee1905_1::eMessageType::ACK_MESSAGE, ieee1905_1::eMessageType::TOPOLOGY_QUERY_MESSAGE,
+            ieee1905_1::eMessageType::CLIENT_CAPABILITY_QUERY_MESSAGE})) {
```
Clang format did something ugly here. Is it possible to fix it?

CoralMalachi replied:
fixed.

morantr replied:
It doesn't seem like that...
Make sure that after you fix it and run the clang_format.sh, it doesn't change it again to the ugly form.

CoralMalachi replied:
I have tried, Travis fails when I change the format by myself.

morantr commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1582,6 +1585,18 @@ bool main_thread::handle_1905_discovery_query(ieee1905_1::CmduMessageRx &cmdu_rx
+ */
+bool main_thread::handle_1905_client_capability_query(ieee1905_1::CmduMessageRx &cmdu_rx)
+{
+    LOG(DEBUG) << "Received Client capability Query message";
```
Ok we received capability query. 
Don't forget to create a response? If you don't know what to fill the TLVs in the response, use dummy values.

tomereli replied:
No need for a response, it is enough to just verify the request was received.

Commit:
travis CI: add client capability test command

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

Commit:
tools: docker: add client capability flow test

Add a test that triggers the sending of a Client Capability Query message
using the WFA_CA format utilized by the WFA_UCC testbed.

The test flow send the command and making sure that the client
capability query was received by the repeater,by checking the logs of
the beerocks_agent_wlan0.

Signed-off-by: Coral Malachi <coral.malachi@intel.com>

CoralMalachi commented:
> Why commit [2a1736a](https://github.com/prplfoundation/prplMesh/commit/2a1736a574853ce50666b8cac5d67921c550d941) is here?

Because the current PR is based on this commit.

tomereli commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -70,6 +70,7 @@ class main_thread : public btl::transport_socket_thread {
                                                 const std::string &src_mac);
     bool handle_1905_discovery_query(ieee1905_1::CmduMessageRx &cmdu_rx);
     //bool sta_handle_event(const std::string &iface,const std::string& event_name, void* event_obj);
+
```
remove accidentally added new line

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4701,8 +4701,9 @@ bool slave_thread::parse_non_intel_join_response(Socket *sd)
+    // LOG(DEBUG) << "Received CLIENT CAPABILITY QUERY MESSAGE";
     const auto mid = cmdu_rx.getMessageId();
-    LOG(DEBUG) << "Received CLIENT_CAPABILITY_QUERY_MESSAGE, mid=" << std::dec << int(mid);
+    LOG(DEBUG) << "Received CLIENT CAPABILITY QUERY MESSAGE , mid=" << std::dec << int(mid);
```
change back

tomereli commented on tools/docker/tests/test_flows.sh:
```diff
@@ -6,7 +6,7 @@
 
 ALL_TESTS="topology initial_ap_config ap_config_renew ap_config_bss_tear_down channel_selection
-           ap_capability_info_reporting client_capability_info_reporting link_metric_collection
+          ap_capability_info_reporting client_capability_info_reporting link_metric_collection
```
why this change of indentation?

tomereli commented on tools/docker/tests/test_flows.sh:
```diff
@@ -27,6 +27,7 @@ start_gw_repeater() {
 }
 
 send_bml_command() {
+    echo docker exec -it gateway ${installdir}/bin/beerocks_cli -c "$*"
```
no need for the echo

tomereli commented on tools/docker/tests/test_flows.sh:
```diff
@@ -98,8 +99,12 @@ test_ap_capability_info_reporting() {
 test_client_capability_info_reporting() {
-    #TODO: Implement
-    return 1
+#TODO: Implement    
```
remove TODO - you're doing this

tomereli commented on tools/docker/tests/test_flows.sh:
```diff
@@ -98,8 +99,12 @@ test_ap_capability_info_reporting() {
-    return 1
+#TODO: Implement    
+    status "test client capability"   
+    eval send_bml_command '"bml_wfa_ca_controller \"DEV_SEND_1905,DestALid,$AL_MAC,MessageTypeValue,0x8009,tlv_type,0x90,tlv_length,0xC,tlv_value,{$RADIO_WLAN0_MAC 0x000000110022}\""' $redirect
```
split to multiple lines

tomereli commented on tools/docker/tests/test_flows.sh:
```diff
@@ -98,8 +99,12 @@ test_ap_capability_info_reporting() {
+    eval send_bml_command '"bml_wfa_ca_controller \"DEV_SEND_1905,DestALid,$AL_MAC,MessageTypeValue,0x8009,tlv_type,0x90,tlv_length,0xC,tlv_value,{$RADIO_WLAN0_MAC 0x000000110022}\""' $redirect
+    sleep 1
+    dbg "Confirming client capability query has been received on agent"
+    docker exec -it repeater sh -c 'grep -i -q "CLIENT CAPABILITY" /tmp/$USER/beerocks/logs/beerocks_agent_wlan0.log'
```
Also, check that the other agent received it. Currently, both will just print `CLIENT CAPABILITY` since they have the same handle, but in the future, we will add a check to verify if the `BSSID == radio mac`.

tomereli commented on tools/docker/tests/test_flows.sh:
```diff
@@ -151,6 +156,8 @@ test_init() {
         exit 1
     }
     AL_MAC=$(docker exec -it gateway ${installdir}/bin/beerocks_cli -c bml_conn_map | grep IRE_BRIDGE | awk '{print $5}' | cut -d ',' -f 1)
+    RADIO_WLAN0_MAC=$(docker exec -it gateway ${installdir}/bin/beerocks_cli -c bml_conn_map | grep "RADIO: wlan0" | head -1 | awk '{print $4}' | cut -d ',' -f 1)
```
Add also `RADIO_WLAN2_MAC`. Might not be used for now, but can be useful for the future.

tomereli commented on tools/docker/tests/test_flows.sh:
```diff
@@ -209,5 +216,6 @@ main() {
 
 VERBOSE=false
 AL_MAC=
+RADIO_WLAN0_MAC=
```
Add RADIO_WLAN0_MAC
